### PR TITLE
Scrapper, Mozgovoy Sergey - 21FPL2

### PR DESCRIPTION
**Pull request by Sergo-M**

### CODE REVIEW

**Comment by Sergo-M** [comment]
Scrapper passes all tests locally (including with num articles = 100). Tests in github are canceled most likely due to the long operation of the program.

**Comment by Sergo-M** [comment]
It seems that there is a problem with connection to econs.online from github. Locally passing all test takes only around 1 minute.

**Comment by Sergo-M** [comment]
Unfortunately, neither adding headers nor adding time.sleep could solve the problem with connecting to the site from github.

File: lab_5_scrapper/scrapper.py [hunk 1]
```diff
@@ -1,164 +1,333 @@
...
-        pass
+        self.config = config
+        self.urls = []
 
     def _extract_url(self, article_bs: BeautifulSoup) -> str:
         """
         Finds and retrieves URL from HTML
         """
-        pass
+        return 'https://econs.online' + str(article_bs.a.get('href'))
```
Comment (dmitry-uraev) [inline lab_5_scrapper/scrapper.py]: Is there another way of extracting url? without hardcoding string into the method?

File: lab_5_scrapper/scrapper.py [hunk 2]
```diff
@@ -1,164 +1,333 @@
...
+    return response
 
 
-class Crawler:
+def is_article(url: str) -> bool:
     """
-    Crawler implementation
+    /photo/ and /tests/ are not articles
     """
+    return '/articles/photo/' not in url and '/tests/' not in url
```
Comment (dmitry-uraev) [inline lab_5_scrapper/scrapper.py]: is it the only way to check article? as a separate function?

File: lab_5_scrapper/scrapper.py [hunk 3]
```diff
@@ -1,164 +1,333 @@
...
-        pass
+        return 'https://econs.online' + str(article_bs.a.get('href'))
 
     def find_articles(self) -> None:
         """
         Finds articles
         """
-        pass
+        count_urls = 0
+        while count_urls < self.config.get_num_articles():
```
Comment (dmitry-uraev) [inline lab_5_scrapper/scrapper.py]: if you get to many articles to collect, how your collection suppose to stop if you can not get all articles requested?

File: lab_5_scrapper/scrapper.py [hunk 4]
```diff
@@ -1,164 +1,333 @@
...
     """
-    pass
+    config = Config(CRAWLER_CONFIG_PATH)
+    prepare_environment(ASSETS_PATH)
+    crawler = Crawler(config=config)
+    crawler.find_articles()
+    for i, url in enumerate(crawler.urls, start=1):
+        parser = HTMLParser(full_url=url, article_id=i, config=config)
+        article = parser.parse()
+        if isinstance(article, Article):
```
Comment (dmitry-uraev) [inline lab_5_scrapper/scrapper.py:325]: why do you specify this condition?

**Review by dmitry-uraev — CHANGES_REQUESTED** [review]
Good start! Please, respond to my comments and/or fix implementation accordingly.

**Comment by dmitry-uraev** [comment]
> Unfortunately, neither adding headers nor adding time.sleep could solve the problem with connecting to the site from github.

We will check on your problem.

File: lab_5_scrapper/scrapper.py (hunk 1 above)
Comment (Sergo-M) [reply lab_5_scrapper/scrapper.py]: > Is there another way of extracting url? without hardcoding string into the method?

Done. Now this is done using urllib.parse.

File: lab_5_scrapper/scrapper.py (hunk 2 above)
Comment (Sergo-M) [reply lab_5_scrapper/scrapper.py]: > is it the only way to check article? as a separate function?

Done. Now this is done using a generator inside the function.

File: lab_5_scrapper/scrapper.py (hunk 3 above)
Comment (Sergo-M) [reply lab_5_scrapper/scrapper.py]: > if you get to many articles to collect, how your collection suppose to stop if you can not get all articles requested?

Done. Now the number of seed_urls is unlimited.

File: lab_5_scrapper/scrapper.py (hunk 4 above)
Comment (Sergo-M) [reply lab_5_scrapper/scrapper.py:325]: > why do you specify this condition?

This is due to the fact that the to_meta() and to_raw() functions take strictly the Article class as input, and the parse() function outputs Union[Article, bool, list]. With this condition, we formally restrict the possible output to the Article class.

File: lab_5_scrapper/scrapper.py (hunk 2 above)
Comment (Sergo-M) [reply lab_5_scrapper/scrapper.py]: > is it the only way to check article? as a separate function?

Due to the target site change, this check is no longer needed.

File: lab_5_scrapper/scrapper.py [hunk 5]
```diff
@@ -1,164 +1,327 @@
...
-        pass
+        self.config = config
+        self.urls = []
 
     def _extract_url(self, article_bs: BeautifulSoup) -> str:
         """
         Finds and retrieves URL from HTML
         """
-        pass
+        base = 'https://kubnews.ru'
```
Comment (dmitry-uraev) [inline lab_5_scrapper/scrapper.py]: can you extract url from each seed_url dynamically by some pattern maybe?

Comment (dmitry-uraev) [reply lab_5_scrapper/scrapper.py]: this is still hardcoding

**Review by dmitry-uraev — COMMENTED** [review]
Nice! Some minor comments only

File: lab_5_scrapper/scrapper.py (hunk 5 above)
Comment (Sergo-M) [reply lab_5_scrapper/scrapper.py]: Now it should do. The base part of the url is now taken from the config (and not specified directly).

**Comment by vitazaytseva** [comment]
Well done! You can proceed to work on Lab 6)

**Pull request closed without merging**